node/src/block_producer/vrf_evaluator/block_producer_vrf_evaluator_reducer.rs

```diff
@@ -1,26 +1,48 @@
+use vrf::VrfEvaluationOutput;
+
+use crate::{
+    block_producer::to_epoch_and_slot,
+    block_producer_effectful::vrf_evaluator_effectful::BlockProducerVrfEvaluatorEffectfulAction,
+    ledger::read::{LedgerReadAction, LedgerReadInitCallback, LedgerReadRequest},
+    BlockProducerAction, Substate,
+};
+
 use super::{
     BlockProducerVrfEvaluatorAction, BlockProducerVrfEvaluatorActionWithMetaRef,
     BlockProducerVrfEvaluatorState, BlockProducerVrfEvaluatorStatus, PendingEvaluation,
-    VrfWonSlotWithHash,
+    SlotPositionInEpoch, VrfWonSlotWithHash,
 };
 
 impl BlockProducerVrfEvaluatorState {
-    pub fn reducer(&mut self, action: BlockProducerVrfEvaluatorActionWithMetaRef<'_>) {
+    pub fn reducer(
+        mut state_context: Substate<Self>,
+        action: BlockProducerVrfEvaluatorActionWithMetaRef<'_>,
+    ) {
+        let Ok(state) = state_context.get_substate_mut() else {
+            return;
+        };
+
         let (action, meta) = action.split();
         match action {
             BlockProducerVrfEvaluatorAction::EvaluateSlot { vrf_input } => {
-                self.status = BlockProducerVrfEvaluatorStatus::SlotEvaluationPending {
+                state.status = BlockProducerVrfEvaluatorStatus::SlotEvaluationPending {
                     time: meta.time(),
                     global_slot: vrf_input.global_slot,
                 };
+
+                state_context.into_dispatcher().push(
```

Comment:
lets `let dispatcher = ...;` for now even if it is just for a single action like everywhere else.

If the case is too common we can add a helper method to the context later.